kod/object/active/holder/room/monsroom/bossroom/ratdungeon.kod

```diff
@@ -0,0 +1,238 @@
+   piOutside_factor = 0
+
+   prMusic = RatDungeon_music
+
+   piGen_time = 3000
+   piGen_percent = 100
+
+   piInit_count_min = 7
+   piInit_count_max = 15
+
+   piMonster_count_max = 70
+
+messages:
+
+   Constructor()
+   {
+      plMonsters = [ [&GiantRat, 100] ];
+      plGenerators = [ [4,38], [5,35], [6,37],
+                       [8,27], [8,30], [8,29],
+                       [18,29], [17,28], [16,28],
+                       [19,22], [18,21], [19,20],
+                       [6,19], [4,19], [5,19],
+                       [2,8], [2,11], [2,10],
+                       [38,15], [36,18], [32,18], [31,14],
+                       [24,16], [28,21], [24,22], [31,27], [27,30]
+                     ];
+
+      propagate;
+   }
+
+   SetupBossTemplates()
+   {
+      plBossTemplate = [ [&GiantRatKing, 36, 41] ];
+      plHenchmenTemplate = [ [&GiantRatSoldier, 33, 40],
+                             [&GiantRatSoldier, 32, 40],
+                             [&GiantRatSoldier, 29, 37],
+                             [&GiantRatSoldier, 27, 39],
+                             [&GiantRatSoldier, 29, 41],
+                             [&GiantRatSoldier, 35, 37],
+                             [&GiantRatSoldier, 37, 39],
+                             [&GiantRatSoldier, 37, 40] ];
+
+      plBossTreasureTemplate = [ [&Gauntlet, 33, 40, 0, 0],
+                                 [&FineLute, 29, 37, 0, 0],
+                                 [&ScaleArmor, 27, 39, 0, 0],
+                                 [&Axe, 35, 37, 0, 0] ,
+                                 [&HealWand, 36, 41, 0, 0]
+                               ];
+
+
+      return;
+   }
+
+   CreateStandardObjects()
+   {
+      local lSkulls;
+
+      lSkulls = [ [13, 36],
+                  [13, 18],
+                  [15, 3],
+                  [36, 8],
+                  [30, 42],
+                  [26, 37],
+                  [32, 36]];
+
+      Send(self,@CreateObjFromList,#lObjlist=lSkulls,
+            #oObjectClassName=&Skull);
+            
+      return;
+   }
+
+
+   CreateStandardExits()
+   {
+      propagate;
+   }
+
+%%%%%%%%%%%%%%%%%%%%%
+% Monster death code%
+%%%%%%%%%%%%%%%%%%%%%
+%
+
+   BossKilledTrigger()
+   {
+      local i, each_obj;
+
+      Send(self,@SomethingWaveRoom,#wave_rsc=RatDungeon_door_up_sound);
+      Send(self,@SetSector,#sector=1,
+            #animation=ANIMATE_FLOOR_LIFT,
+            #height=770,#speed=20);
+      Send(self,@SetSector,#sector=2,
+            #animation=ANIMATE_FLOOR_LIFT,
+            #height=770,#speed=20);
+      Send(self,@SetSector,#sector=3,
+            #animation=ANIMATE_FLOOR_LIFT,
+            #height=770,#speed=20);
+      
+      pbGenerateMonsters = FALSE;
+      
+      foreach i in plActive
+      {
+         each_obj = Send(self,@HolderExtractObject,#data=i);
+         if IsClass(each_obj,&GiantRat)
+            OR IsClass(each_obj,&GiantRatSoldier)
+         {
+            Send(each_obj,@Delete);
```

Comment:
I thought about that, but the rats haven't died, they've simply 'run away' - and they are in areas inaccessible to the boss pit. Not sure if there's a way to do it better. Do we have actual "Run into holes" (and then despawn) AI? That would be awesome